cpp/spglib.hpp

```diff
@@ -0,0 +1,101 @@
+#pragma once
+
+#include <array>
+#include <exception>
+#include <memory>
+#include <string>
+#include <tuple>
+#include <vector>
+
+namespace Spglib {
+	/**
+	 * Get the version of the Spglib C++ binding
+	 *
+	 * @return X.Y.Z version
+	 */
+	const std::string version();
+	/**
+	 * Get the full version of the Spglib C++ binding
+	 *
+	 * @return full version
+	 */
+	const std::string version_full();
+	/**
+	 * Get the commit of the Spglib C++ binding
+	 *
+	 * @return commit
+	 */
+	const std::string commit();
+	/**
+	 * C++ exception object for spglib related errors
+	 */
+	class SpglibError : public std::exception {
+		const std::string errorMsg;
+
+	public:
+		const char* what() const noexcept override;
+		SpglibError(std::string&& message);
+	};
+	class Symmetry {
+		// TODO: Find better interface compatible with std::mdspan
+		std::array<double, 9> rotations;
+		std::array<double, 3> translations;
+	};
+	class MagneticSymmetry : public Symmetry {
+		bool time_reversal;
+	};
+	class Atom {
+		std::array<double, 3> position;
+		int type;
+	};
+	// TODO: Better name?
+	class Lattice {
+		std::array<double, 9> lattice;
+		std::vector<Atom> atoms;
+		// TODO: Allow for non-periodic dimensions
+	};
+	class PrimitiveLattice : public Lattice {
+		Lattice& original_lattice;
+	};
+	class Spacegroup {
+		int number;
+		int hall_number;
+		std::string international_symbol;
+		std::string hall_symbol;
+		std::string schoenflies;
+		std::string pointgroup_international;
+		std::string pointgroup_schoenflies;
+		int arithmetic_crystal_class_number;
+		std::string arithmetic_crystal_class_symbol;
+		// TODO: Unclear what this is referencing
+		char choice[6];
+	};
+	// TODO: should it inherit/reference Spacegroup
+	class MagneticSpacegroup {
+		int number;
+		int type;
+		int uni_number;
+		int litvin_number;
+		std::string bns_number;
+		std::string og_number;
+	};
+	// TODO: Combine/templetize
+	class Dataset {
+		std::vector<Symmetry> symmetries;
+		Spacegroup spacegroup;
+		// These are dependent on the input, should be moved to constructor?
+		std::array<double, 9> transformation_matrix;
+		std::array<double, 3> origin_shift;
+		// Smart reference to the input used
+		std::shared_ptr<Lattice> lattice;
+	};
+	class MagneticDataset {
+		std::vector<MagneticSymmetry> symmetries;
+		MagneticSpacegroup spacegroup;
+		// These are dependent on the input, should be moved to constructor?
+		std::array<double, 9> transformation_matrix;
+		std::array<double, 3> origin_shift;
+		// Smart reference to the input used
+		std::shared_ptr<Lattice> lattice;
+	};
```

Comment:
I think for these we only need to store `symmetries`, `spacegroup` and `lattice`, what do you think?

Reply:
`SpglibDataset` provides multiple different kinds of information.
- Matrix representations of space group operations for input crystal structure
- Space group type including setting, choice, point group, crystal class, that are covered by `SpglibSpacegroupType` and its associated information of the input crystal structure, Wyckoff positions, etc.
- Crystal structure that standardizes the input crystal structure
- Transformation matrix (P, p) from the input structure to the conventional unit cell
- Primitive cell

Depending on purposes, most of these information are essential. All these were squashed in to the `SpglibDataset` just for convenience. Primitive cell information may not be necessary, although it was necessary at some stage in history.

Reply:
Indeed and most of them will be made available indirectly via getters. The dataset itself should only store the minimum distinguishable information.

E.g. transformation matrix can be derived from `lattice` and `std_lattice`/`primitive_lattice`/`spacegrouptype`

`MagbeticDataset` I am definitely missing stuff there.

So far it seems I should include `std_lattice` and `primitive_lattice `

As for interface, I should add:
- transformation matrices
- wyckoff positions
- equivalent atoms

Other things should be taken indirectly, e.g. `hall_number` through `spacegroup` (should be `spacegroup_type` here?)

Reply:
I think the transformation matrix is the most important information.

In crystallography, we usually don't think about Cartesian coordinates. However, `std_lattice` contains (1) change-of-basis (transformation) matrix, (2) rigid rotation in Cartesian coordinates, (3) removal of distortion. So in addition, `std_rotation_matrix` is necessary to relate them. `primitive_lattice` is an intermediate information and it is not standardized. Probably this information is rarely used. To make primitive cell, we can transform the input crystal structure using the transformation matrix, then we can apply a pre-defined transformation matrix to the primitive cell, https://spglib.readthedocs.io/en/stable/definition.html#transformation-to-the-primitive-cell. This is a cleaner approach.

The strong point of spglib is the feature of removing distortion. The way of the treatment of distortion is not unique, and is somehow implemented algorithmically not mathematically in spglib. These naive information are averaged out and pushed into each element of the dataset. Spglib pretends working properly, but no one knows its super details.